source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -1224,20 +1221,57 @@ void Graph::setUiNodeInfo(UiNodePtr node, const std::string& type, const std::st
 void Graph::createNodeUIList(mx::DocumentPtr doc)
 {
     _nodesToAdd.clear();
-    const std::string EXTRA_GROUP_NAME = "extra";
-    for (mx::NodeDefPtr nodeDef : doc->getNodeDefs())
+
+    std::vector<std::string> ordered_groups = {
+        "texture2d",
+        "texture3d",
+        "procedural",
+        "procedural2d",
+        "procedural3d",
+        "geometric",
+        "translation",
+        "convolution2d",
+        "math",
+        "adjustment",
+        "compositing",
+        "conditional",
+        "channel",
+        "organization",
+        "global",
+        "application",
+        "material",
+        "shader",
+        "pbr",
+        "light",
+        "colortransform"
+    };
+
+    auto nodeDefs = doc->getNodeDefs();
+    std::unordered_map<std::string, std::vector<mx::NodeDefPtr>> groupToNodeDef;
+
+    for (const auto& nodeDef : nodeDefs)
     {
-        // NodeDef is the key for the map
         std::string group = nodeDef->getNodeGroup();
-        if (group.empty())
+
+        if (groupToNodeDef.find(group) == groupToNodeDef.end())
```

Comment:
Want to check here, if there is no node group (empty string), what will occur ? Is it better to create some string name for this group ? (While this was cleaned up in the std library, there is no requirement to have a nodegroup and I've seen in DCCs where it's not set).

Reply:
Based on my understanding, these would not get added to the menu at all. To me that seems like a bug. @lfl-eholthouser would it make sense to put all nodes that are not in the hard coded list of groups at the bottom of the menu?

Reply:
Yes, I would be interested in knowing when nodedefs would not have a NodeGroup. But I think that would definitely make sense maybe in something like a Misc group? Just so as not to greatly increase the size of the main menu

Reply:
I've added `no_group` as an option for those that do not have a group

Reply:
Thanks @beersandrew. I leave it up to others on the name of "no_group".  Maybe "Ungrouped" instead as it shows up in the UI. Also you might want to take care in case someone calls their group "no_group" :).